app/monitoringapi.go

```diff
@@ -0,0 +1,105 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package app
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	eth2client "github.com/attestantio/go-eth2-client"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p/p2p/protocol/ping"
+
+	"github.com/obolnetwork/charon/app/errors"
+)
+
+// newReadyHandler returns a http.HandlerFunc which returns 200 when both the beacon node is synced and all quorum peers can be pinged  in parallel within a timeout. Returns 500 otherwise.
+func newReadyHandler(ctx context.Context, tcpNode host.Host, eth2Cl eth2client.NodeSyncingProvider, peerIDs []peer.ID) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		syncing, err := beaconNodeSyncing(ctx, eth2Cl)
+		if err != nil {
+			writeResponse(w, http.StatusInternalServerError, "Failed to get beacon sync state")
+		}
+
+		if syncing {
+			writeResponse(w, http.StatusInternalServerError, "Beacon node not synced")
+			return
+		}
+
+		err = peersReady(ctx, peerIDs, tcpNode)
+		if err != nil {
+			writeResponse(w, http.StatusInternalServerError, "Couldn't ping all peers")
+			return
+		}
+
+		writeResponse(w, http.StatusOK, "ok")
+	}
+}
+
+// beaconNodeSyncing returns true if the beacon node is still syncing.
+func beaconNodeSyncing(ctx context.Context, eth2Cl eth2client.NodeSyncingProvider) (bool, error) {
+	state, err := eth2Cl.NodeSyncing(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	return state.IsSyncing, nil
+}
+
+// peersReady returns nil if all quorum peers can be pinged in parallel within a timeout. Returns error otherwise.
+func peersReady(ctx context.Context, peerIDs []peer.ID, tcpNode host.Host) error {
```

Comment:
you need to cancel the context when returning to stop the long lived ping connections:
```
ctx, cancel := context.WithCancel(ctx)
defer cancel()
```

Reply:
makes sense!